### PR TITLE
V2: training with cross validation

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This PR intends to resolve some parts of https://github.com/chemprop/chemprop/issues/524 and https://github.com/chemprop/chemprop/issues/581.

- [x] Fix the issue where the same list of indices repeat num_folds times
- [x] The `data_split` function is not doing cross-validation correctly. See https://github.com/chemprop/chemprop/issues/609. Needs it to be fixed to complete this PR.
- [x] Add support to use cross validation in the training script
- [x] Add support to use ensembling in the training script
- [x] Add training with cross validation in cli test

Below are some relevant discussions:
- Output directory structure for cross-validation and ensembling: https://github.com/chemprop/chemprop/issues/581#issuecomment-1896322966 (may need to consult https://github.com/chemprop/chemprop/issues/591, and https://github.com/chemprop/chemprop/pull/599 for the output structure and location)

### CODE REVIEW

**Comment by davidegraff** [comment]
can we break this up into two PRs: (i) ensembling and then (ii) cross-validation? The former will fundamentally change the object model of the train and predict CLIs while the latter will _sometimes_ change it? These are both decently large features and should be reviewed independently.

**Comment by hwpang** [comment]
@davidegraff Thanks for the comment! Yeah that makes sense. I will split up this PR.

**Comment by hwpang** [comment]
@davidegraff I have removed the changes relevant to ensembling from this PR, and this PR should now only contain cross-validation related efforts. I moved the ensembling related efforts to https://github.com/chemprop/chemprop/pull/631. The ensembling efforts depend on this PR, so I think it's easier to review and merge this one in, and then move onto https://github.com/chemprop/chemprop/pull/631.

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
-                train_data, val_data, _ = split_multicomponent(all_data, args.split, **split_kwargs)
-            else:
-                train_data, val_data, _ = split_monocomponent(all_data, args.split, **split_kwargs)
-    else:
+            needs_val_data = False
+    elif args.separate_val_path is not None:
         raise ArgumentError(
-            argument=None, message="'val_path' must be specified if 'test_path' is provided!"
+            argument=None, message="'test_path' must be specified if 'val_path' is provided!"
         )  # TODO: In v1 this wasn't the case?
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:538]: This TODO might need to be broken out into a separate issue?

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+                if args.task_type == "regression":
+                    model.predictor.register_buffer("loc", torch.tensor(scaler.mean_).view(-1, 1))
+                    model.predictor.register_buffer(
+                        "scale", torch.tensor(scaler.scale_).view(-1, 1)
+                    )
+                results = trainer.test(model, test_loader)[0]
+                logger.info(f"Test results: {results}")
+
+            p_model = args.output_dir / "model.pt"
+            input_scalers = []  # TODO: we should add descriptor scalers here
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:744]: Again, consider breaking this TODO into an issue

Comment (JacksonBurns) [reply chemprop/cli/train.py:744]: (and removing the inline comment)

File: chemprop/cli/train.py [hunk 3]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+                    for i in range(n_components)
+                ]
+                if args.mpn_shared:
+                    mp_block = MulticomponentMessagePassing(
+                        mp_blocks[0], n_components, args.mpn_shared
+                    )
+                else:
+                    mp_block = MulticomponentMessagePassing(
+                        mp_blocks, n_components, args.mpn_shared
+                    )
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:612]:
```
mp_block = MulticomponentMessagePassing(
    mp_blocks[0] if args.mpn_shared else mp_blocks, n_components, args.mpn_shared
)
```

File: chemprop/cli/train.py [hunk 4]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+
+            if multicomponent:
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:680]: This can be moved outside of the `if multicomponent:` block right? (and removed from both places in here).

File: chemprop/cli/train.py [hunk 5]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+            logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+
+            if no_cv:
+                output_dir = args.output_dir
+            else:
+                output_dir = args.output_dir / f"fold_{fold_idx}"
+
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:715]: Super nitpick, but I don't see a reason to abbreviate this (even though this is close to the file extension) when `checkpoints` would be more clear.

File: chemprop/cli/train.py [hunk 6]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+                "{epoch}-{val_loss:.2f}",
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:716]: Is it possible for this to overwrite previous checkpoints?

File: chemprop/cli/train.py [hunk 7]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+                "val_loss",
+                mode=monitor_mode,
+                save_last=True,
+            )
+
+            early_stopping = EarlyStopping("val_loss", patience=5, mode=monitor_mode)
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:722]: RE: our discussion during the hackathon last week. v1 would never actually _stop_ training, it would just load the best model from all epochs, acting as if it had stopped there. This will be different, in that v2 will actually _stop training_. Should we make this _very_ clear somewhere? Is this intended?

File: chemprop/cli/train.py [hunk 8]
```diff
@@ -542,170 +531,220 @@ def main(args):
...
+
+            trainer = pl.Trainer(
+                logger=tb_logger,
+                enable_progress_bar=True,
+                accelerator="auto",
+                devices=args.n_gpu if torch.cuda.is_available() else 1,
```
Comment (JacksonBurns) [inline chemprop/cli/train.py:728]: There's a bit of difficulty in this - lightning has a lot of ways to use `devices`, see: https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#:~:text=The%20table%20below%20lists%20examples%20of%20possible%20input%20formats%20and%20how%20they%20are%20interpreted%20by%20Lightning.

We should follow up about this offline. The way we pass this arg should probably change if we want to expose all the functionality of `devices`.

**Review by JacksonBurns — COMMENTED** [review]
Some minor details, but some larger questions for discussion. Overall, this looks quite good.

File: chemprop/cli/train.py (hunk 3 above)
Comment (davidegraff) [reply chemprop/cli/train.py:612]: Is this `if/else` necessary? It seems like it's taken care of inside [`MulticomponentMessagePassing.__init__`](https://github.com/chemprop/chemprop/blob/327c31e932db8cea92e9d9a221c288eaeee911cb/chemprop/nn/message_passing/multi.py#L50):
```python
self.blocks = nn.ModuleList([blocks[0]] * self.n_components if shared else blocks)
```

File: chemprop/cli/train.py (hunk 6 above)
Comment (davidegraff) [reply chemprop/cli/train.py:716]: yes, provided they have the `output_dir` and same validation loss (to 2 decimal points) on the same epoch

File: chemprop/cli/train.py (hunk 7 above)
Comment (davidegraff) [reply chemprop/cli/train.py:722]: We should make this configurable. I think I just added this in when I originally scaffolded out the training CLI from other code I've written. Long term, it would be nice if we could have a separate config file for callbacks, but for now, we can just add a `--patience` arg and set the default to `None` to indicate training for the full number of epochs:
```python
patience = args.patience if args.patience is not None else args.epochs
```

File: chemprop/cli/train.py (hunk 8 above)
Comment (davidegraff) [reply chemprop/cli/train.py:728]: Agreed. IMO we should have an `int` arg `--devices` that's just the _number_ of devices to train on. If a user wants to limit the GPU set, they can just run it like so:
```
$ CUDA_VISIBLE_DEVICES=0,2 chemprop train --devices -1
```
or we could also use the `find_usable_cuda_devices` method so that a user doesn't even need to bother with limiting GPU visibility

File: chemprop/cli/train.py (hunk 1 above)
Comment (hwpang) [reply chemprop/cli/train.py:538]: Thanks! I made this issue to keep track of this TODO: https://github.com/chemprop/chemprop/issues/639.

File: chemprop/cli/train.py (hunk 2 above)
Comment (hwpang) [reply chemprop/cli/train.py:744]: Added this issue, will remove the inline comment: https://github.com/chemprop/chemprop/issues/640

File: chemprop/cli/train.py (hunk 3 above)
Comment (hwpang) [reply chemprop/cli/train.py:612]: @davidegraff Good call! You're right. I will remove this `if/else`.

File: chemprop/cli/train.py (hunk 4 above)
Comment (hwpang) [reply chemprop/cli/train.py]: Thanks for the suggestion! I move this outside of the `if multicomponent:` block.

File: chemprop/cli/train.py (hunk 6 above)
Comment (hwpang) [reply chemprop/cli/train.py:708]: With changes implemented in https://github.com/chemprop/chemprop/pull/599, I think `output_dir` would be `$PWD/chemprop_training/<input_stem>/$NOW` so this will not be overwritten.

File: chemprop/cli/train.py (hunk 5 above)
Comment (hwpang) [reply chemprop/cli/train.py:707]: I think this is outside the scope of this PR as I only move this part of code into the for loop and did not change the directory name. We can open a different issue to discuss whether to abbreviate output directories.

File: chemprop/cli/train.py (hunk 7 above)
Comment (hwpang) [reply chemprop/cli/train.py]: Thanks for the suggestions! I have added a `--patience` arg and set the default to `None` as suggested.

File: chemprop/data/splitting.py [hunk 9]
```diff
@@ -87,7 +87,7 @@ def split_data(
     train, val, test = None, None, None
     match SplitType.get(split):
         case SplitType.CV_NO_VAL | SplitType.CV:
-            if (max_folds := len(datapoints)) < num_folds or num_folds <= 1:
+            if not (2 <= num_folds <= len(datapoints)):
                 raise ValueError(
                     f"Number of folds for cross-validation must be between 3 and {max_folds} (length of data) inclusive for `--split cv` and 2 and {max_folds} (length of data) inclusive for `--split cv_no_val` (got {num_folds})."
                 )
```
Comment (davidegraff) [inline chemprop/data/splitting.py:93]:
```suggestion
                min_folds = 2 if SplitType.get(split) == SplitType.CV_NO_VAL else 3
                raise ValueError(
                    f"invalid number of folds requested! got: {num_folds}, but expected between {min_folds} and {len(datapoints)} (i.e., number of datapoints), inclusive, for split type: {repr(split)}"
                )
```

File: chemprop/cli/train.py [hunk 10]
```diff
@@ -458,6 +458,12 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         default=0,
         help="Seed for PyTorch randomness (e.g., random initial weights).",
     )
+    parser.add_argument(
+        "--patience",
+        type=int,
+        default=None,
+        help="Number of epochs to wait for improvement before early stopping.",
+    )
```
Comment (davidegraff) [inline chemprop/cli/train.py:466]: Note: this can stay because we've discussed this already, but stuff like this should be a separate PR in the future (there really is no minimum size!)

Comment (hwpang) [reply chemprop/cli/train.py:466]: Thanks for the suggestion! Will keep this change here and open up new PRs in the future.

**Comment by davidegraff** [comment]
> Just so I'm clear on the logical structure here, you've restructured the train CLI to effectively treat _all_ training as having `k` separate splits and training one model for each of those splits. It's at the end of model training that you then return and check whether this model was trained as part of CV or not, and if so, structure the output directory as `OUTPUT/FOLD_K/...`. Otherwise, you structure it normally? I just want to make sure I understand this before I approve. Thanks!

@hwpang can you check if this is right?

**Comment by hwpang** [comment]
@davidegraff Yes that is correct!

File: chemprop/cli/train.py [hunk 11]
```diff
@@ -691,63 +701,83 @@ def main(args):
...
+    no_cv = args.num_folds == 1
+    if no_cv:
+        train_datas = [train_data]
+        val_datas = [val_data]
+        test_datas = [test_data]
     else:
-        test_loader = None
+        train_datas = train_data
+        val_datas = val_data
+        test_datas = test_data
```
Comment (davidegraff) [inline chemprop/cli/train.py:710]:
```suggestion
        splits = (train_data, val_data, test_data)
```

File: chemprop/cli/train.py [hunk 12]
```diff
@@ -691,63 +701,83 @@ def main(args):
...
-    val_loader = MolGraphDataLoader(val_dset, args.batch_size, args.num_workers, shuffle=False)
-    if test_dset is not None:
-        test_loader = MolGraphDataLoader(
-            test_dset, args.batch_size, args.num_workers, shuffle=False
-        )
+    no_cv = args.num_folds == 1
+    if no_cv:
+        train_datas = [train_data]
+        val_datas = [val_data]
+        test_datas = [test_data]
```
Comment (davidegraff) [inline chemprop/cli/train.py:706]:
```suggestion
        splits = ([train_data], [val_data], [test_data])
```

File: chemprop/cli/train.py [hunk 13]
```diff
@@ -691,63 +701,83 @@ def main(args):
...
     else:
-        test_loader = None
+        train_datas = train_data
+        val_datas = val_data
+        test_datas = test_data
 
-    model = build_model(args, train_dset)
-    logger.info(model)
+    for fold_idx, (train_data, val_data, test_data) in enumerate(zip(train_datas, val_datas, test_datas)):
```
Comment (davidegraff) [inline chemprop/cli/train.py:712]:
```suggestion
    for fold_idx, (train_data, val_data, test_data) in enumerate(splits):
```

**Review by davidegraff — APPROVED** [review]
minor style suggestions, otherwise lgtm

**Comment by hwpang** [comment]
All tests pass! I believe this is now ready to merge.

**Review by JacksonBurns — APPROVED** [review]

File: chemprop/cli/train.py (hunk 5 above)
Comment (KnathanM) [reply chemprop/cli/train.py]: I agreed with Jackson and included this in my PR #599
